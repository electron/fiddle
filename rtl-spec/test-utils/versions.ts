import {
  ElectronReleaseChannel,
  InstallState,
  VersionSource,
} from '../../src/interfaces';
import { VersionsMock } from '../../tests/mocks/electron-versions';
import { StateMock } from '../../tests/mocks/state';

const { missing } = InstallState;
const { remote } = VersionSource;

export const mockVersion1 = {
  source: remote,
  state: missing,
  version: '26.0.0',
};

export const mockVersion2 = {
  source: remote,
  state: missing,
  version: '28.0.0-unsupported',
};

/**
 * Initializes the app state with our mock versions.
 */
export function prepareAppState() {
  const { state: appState } = window.app;

  const { mockVersions } = new VersionsMock();

  (appState as unknown as StateMock).initVersions('2.0.2', {
    ...mockVersions,
    [mockVersion1.version]: { ...mockVersion1 },
    [mockVersion2.version]: { ...mockVersion2 },
  });

  appState.channelsToShow = [
    ElectronReleaseChannel.stable,
    ElectronReleaseChannel.beta,
  ];

  return appState;
}
