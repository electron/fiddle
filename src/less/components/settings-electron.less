@import '../variables.less';

.settings-electron {
  .version-filter {
    margin: 10px 0;
  }

  .electron-versions-table {
    border: 1px solid @border-color-1;
    border-radius: 3px;
    overflow: hidden;
    background-color: @background-1;

    .electron-versions-header {
      display: flex;
      font-weight: 600;
      padding: 10px 0;
      border-bottom: 2px solid @border-color-1;
      background-color: @background-2;
      position: sticky;
      top: 0;
      z-index: 10;

      .version-col {
        flex: 1;
        padding: 0 15px;
      }

      .status-col {
        width: 180px;
        padding: 0 15px;
      }

      .action-col {
        width: 160px;
        padding: 0 15px;
        text-align: center;
      }
    }

    .no-versions {
      padding: 40px 20px;
      text-align: center;
      color: @text-color-2;
    }

    .electron-versions-list {
      .electron-version-row {
        display: flex;
        align-items: center;
        border-bottom: 1px solid @border-color-2;

        &.even {
          background-color: @background-1;
        }

        &.odd {
          background-color: rgba(92, 112, 128, 0.05);
        }

        &:hover {
          background-color: rgba(92, 112, 128, 0.15);
        }

        .version-col {
          flex: 1;
          padding: 0 15px;
        }

        .status-col {
          width: 180px;
          padding: 0 15px;

          .bp3-icon {
            margin-right: 5px;
          }
        }

        .action-col {
          width: 160px;
          padding: 0 15px;
          display: flex;
          justify-content: center;

          .bp3-button {
            min-width: 80px;
            max-width: 130px;
          }

          .disabled-version {
            min-width: 80px;
            max-width: 130px;
          }
        }
      }
    }
  }

  // Dark theme overrides
  .bp3-dark & {
    .electron-versions-table {
      .no-versions {
        color: @foreground-1;
      }

      .electron-versions-header {
        background-color: @dark-gray5;
        border-bottom-color: @dark-gray4;
      }

      .electron-versions-list {
        .electron-version-row {
          border-bottom-color: @dark-gray4;

          &.odd {
            background-color: rgba(255, 255, 255, 0.02);
          }

          &:hover {
            background-color: rgba(255, 255, 255, 0.05);
          }
        }
      }
    }
  }
}
