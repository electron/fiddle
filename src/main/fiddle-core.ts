import { ChildProcess } from 'node:child_process';

import { ElectronVersions, Installer, Runner } from '@electron/fiddle-core';
import { BrowserWindow, IpcMainEvent, WebContents } from 'electron';

import { ELECTRON_DOWNLOAD_PATH, ELECTRON_INSTALL_PATH } from './constants';
import { ipcMainManager } from './ipc';
import {
  DownloadVersionParams,
  ProgressObject,
  StartFiddleParams,
} from '../interfaces';
import { IpcEvents } from '../ipc-events';

let installer: Installer;
let runner: Runner;

// Keep track of which fiddle process belongs to which WebContents
const fiddleProcesses = new WeakMap<WebContents, ChildProcess>();

const downloadingVersions = new Map<string, Promise<any>>();
const removingVersions = new Map<string, Promise<void>>();

/**
 * Start running an Electron fiddle.
 */
export async function startFiddle(
  webContents: WebContents,
  params: StartFiddleParams,
): Promise<void> {
  const { dir, env, isValidBuild, localPath, options, version } = params;
  const child = await runner.spawn(
    isValidBuild && localPath ? Installer.getExecPath(localPath) : version,
    dir,
    { args: options, cwd: dir, env },
  );
  fiddleProcesses.set(webContents, child);

  const pushOutput = (data: string | Buffer) => {
    ipcMainManager.send(
      IpcEvents.FIDDLE_RUNNER_OUTPUT,
      [data.toString()],
      webContents,
    );
  };

  child.stdout?.on('data', pushOutput);
  child.stderr?.on('data', pushOutput);

  child.on('close', async (code, signal) => {
    fiddleProcesses.delete(webContents);

    ipcMainManager.send(IpcEvents.FIDDLE_STOPPED, [code, signal], webContents);
  });
}

/**
 * Stop a currently running Electron fiddle.
 */
export function stopFiddle(webContents: WebContents): void {
  const child = fiddleProcesses.get(webContents);
  child?.kill();

  if (child) {
    // If the child process is still alive 1 second after we've
    // attempted to kill it by normal means, kill it forcefully.
    setTimeout(() => {
      if (child.exitCode === null) {
        child.kill('SIGKILL');
      }
    }, 1000);
  }
}

export async function setupFiddleCore(versions: ElectronVersions) {
  // For managing downloads and versions for electron
  installer = new Installer({
    electronDownloads: ELECTRON_DOWNLOAD_PATH,
    electronInstall: ELECTRON_INSTALL_PATH,
  });

  // Broadcast state changes to all windows
  installer.on('state-changed', (event) => {
    for (const window of BrowserWindow.getAllWindows()) {
      ipcMainManager.send(
        IpcEvents.VERSION_STATE_CHANGED,
        [event],
        window.webContents,
      );
    }
  });

  runner = await Runner.create({ installer, versions });

  ipcMainManager.on(
    IpcEvents.GET_VERSION_STATE,
    (event: IpcMainEvent, version: string) => {
      event.returnValue = installer.state(version);
    },
  );
  ipcMainManager.handle(
    IpcEvents.DOWNLOAD_VERSION,
    async (
      event: IpcMainEvent,
      version: string,
      opts?: Partial<DownloadVersionParams>,
    ) => {
      const webContents = event.sender;

      if (removingVersions.has(version)) {
        throw new Error('Version is being removed');
      }

      if (!downloadingVersions.has(version)) {
        const promise = installer.ensureDownloaded(version, {
          ...opts,
          progressCallback: (progress: ProgressObject) => {
            ipcMainManager.send(
              IpcEvents.VERSION_DOWNLOAD_PROGRESS,
              [version, progress],
              webContents,
            );
          },
        });

        downloadingVersions.set(version, promise);
      }

      try {
        await downloadingVersions.get(version);
      } finally {
        downloadingVersions.delete(version);
      }
    },
  );
  ipcMainManager.handle(
    IpcEvents.REMOVE_VERSION,
    async (_: IpcMainEvent, version: string) => {
      if (downloadingVersions.has(version)) {
        throw new Error('Version is being downloaded');
      }

      if (!removingVersions.has(version)) {
        removingVersions.set(version, installer.remove(version));
      }

      try {
        await removingVersions.get(version);
        return installer.state(version);
      } finally {
        removingVersions.delete(version);
      }
    },
  );
  ipcMainManager.handle(
    IpcEvents.START_FIDDLE,
    async (event: IpcMainEvent, params: StartFiddleParams) => {
      await startFiddle(event.sender, params);
    },
  );
  ipcMainManager.on(IpcEvents.STOP_FIDDLE, (event: IpcMainEvent) => {
    stopFiddle(event.sender);
  });
}
