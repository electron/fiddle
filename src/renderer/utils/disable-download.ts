import semver from 'semver';

/**
 * disables download button for versions:
 * - below 11.0.0 on Apple Silicon.
 * - below 6.0.8 and 7.0.0 on Windows arm64
 * Reference: {@link https://www.electronjs.org/blog/apple-silicon}
 *
 * @param version - electron version
 */
export function disableDownload(version: string): boolean {
  return (
    (window.ElectronFiddle.platform === 'darwin' &&
      window.ElectronFiddle.arch === 'arm64' &&
      semver.lt(version, '11.0.0')) ||
    (window.ElectronFiddle.platform === 'win32' &&
      window.ElectronFiddle.arch === 'arm64' &&
      !semver.satisfies(version, '>=6.0.8 || >=7.0.0'))
  );
}
