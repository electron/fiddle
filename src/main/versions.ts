import * as path from 'node:path';

import { ElectronVersions, Installer, SemVer } from '@electron/fiddle-core';
import { IpcMainEvent, app } from 'electron';
import * as fs from 'fs-extra';

import releases from '../../static/releases.json';
import { InstallState, Version } from '../interfaces';
import { IpcEvents } from '../ipc-events';
import { ipcMainManager } from './ipc';

let knownVersions: ElectronVersions;

/**
 * Helper to check if this version is from a released major branch.
 *
 * This way when we have a local version of Electron like '999.0.0'
 * we'll know to not try & download 999-x-y.zip from GitHub :D
 *
 * @param {number} major - Electron major version number
 * @returns {boolean} true if there are releases with that major version
 */
export function isReleasedMajor(major: number): boolean {
  return knownVersions.inMajor(major).length > 0;
}

export function getOldestSupportedMajor(): number | undefined {
  const NUM_BRANCHES = parseInt(process.env.NUM_STABLE_BRANCHES || '');

  if (!Number.isNaN(NUM_BRANCHES)) {
    return knownVersions.stableMajors.slice(-NUM_BRANCHES)[0];
  }

  return knownVersions.supportedMajors[0];
}

export function getLatestStable(): SemVer | undefined {
  return knownVersions.latestStable;
}

export function getReleasedVersions(): Array<Version> {
  // Don't support anything older than 0.30 (Aug 2015).
  // The oldest version known to releases.json.org is 0.20,
  // Pre-0.24.0 versions were technically 'atom-shell' and cannot
  // be downloaded with @electron/get.
  // TODO(dsanders11): upstream this logic to @electron/fiddle-core
  return knownVersions.versions
    .filter((ver) => !ver.version.startsWith('0.2'))
    .map(({ version }) => ({ version }));
}

/**
 * Gets the current state of a specific version
 * Valid local electron builds are marked as `installed`
 *
 * @param {Version} ver
 * @returns {InstallState}
 */
export function getLocalVersionState(ver: Version): InstallState {
  const { localPath } = ver;
  if (localPath !== undefined) {
    const dir = Installer.getExecPath(localPath);
    if (fs.existsSync(dir)) {
      return InstallState.installed;
    }
  }

  return InstallState.missing;
}

export async function fetchVersions(): Promise<Version[]> {
  await knownVersions.fetch();
  return getReleasedVersions();
}

export async function setupVersions() {
  knownVersions = await ElectronVersions.create(
    {
      versionsCache: path.join(app.getPath('userData'), 'releases.json'),
    },
    {
      initialVersions: releases,
    },
  );

  ipcMainManager.handle(
    IpcEvents.IS_RELEASED_MAJOR,
    (_: IpcMainEvent, version: number) => isReleasedMajor(version),
  );
  ipcMainManager.handle(IpcEvents.FETCH_VERSIONS, (_: IpcMainEvent) =>
    fetchVersions(),
  );
  ipcMainManager.on(IpcEvents.GET_LATEST_STABLE, (event) => {
    event.returnValue = getLatestStable();
  });
  ipcMainManager.on(IpcEvents.GET_LOCAL_VERSION_STATE, (event, ver) => {
    event.returnValue = getLocalVersionState(ver);
  });
  ipcMainManager.on(IpcEvents.GET_OLDEST_SUPPORTED_MAJOR, (event) => {
    event.returnValue = getOldestSupportedMajor();
  });
  ipcMainManager.on(IpcEvents.GET_RELEASED_VERSIONS, (event) => {
    event.returnValue = getReleasedVersions();
  });
  ipcMainManager.handle(
    IpcEvents.GET_RELEASE_INFO,
    (_: IpcMainEvent, version) => knownVersions.getReleaseInfo(version),
  );

  return knownVersions;
}
