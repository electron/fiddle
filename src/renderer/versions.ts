import { InstallState, Installer } from '@vertedinde/fiddle-core';
import * as fs from 'fs-extra';
import semver from 'semver';

import releasesJSON from '../../static/releases.json';
import {
  ElectronReleaseChannel,
  RunnableVersion,
  Version,
  VersionSource,
} from '../interfaces';
import { normalizeVersion } from '../utils/normalize-version';

/**
 * Returns a sensible default version string.
 *
 * @param {Array<RunnableVersion>} knownVersions
 * @returns {string}
 */
export function getDefaultVersion(versions: RunnableVersion[]): string {
  const key = localStorage.getItem('version');
  if (key && versions.some(({ version }) => version === key)) return key;

  // newest stable release
  const latestStable = versions
    .filter((ver) => !ver.version.includes('-')) // stable
    .map((ver) => semver.parse(ver.version))
    .sort((a, b) => -semver.compare(a!, b!))
    .shift();
  if (latestStable) return latestStable.version;

  // how do we not have a stable version listed?
  throw new Error('Corrupted version data');
}

/**
 * Return the release channel for a given input
 * version.
 *
 * @param {Version | string} input
 * @returns {ElectronReleaseChannel}
 */
export function getReleaseChannel(
  input: Version | string,
): ElectronReleaseChannel {
  const tag = typeof input === 'string' ? input : input.version || '';

  if (tag.includes('beta') || tag.includes('alpha')) {
    return ElectronReleaseChannel.beta;
  }

  if (tag.includes('nightly')) {
    return ElectronReleaseChannel.nightly;
  }

  // Must be a stable version, right?
  return ElectronReleaseChannel.stable;
}

export const enum VersionKeys {
  local = 'local-electron-versions',
  known = 'known-electron-versions',
}

/**
 * Retrieve Electron versions from localStorage.
 *
 * @param {VersionKeys} key
 * @param {() => Array<Version>} fallbackMethod
 * @returns {Array<Version>}
 */
function getVersions(
  key: VersionKeys,
  fallbackMethod: () => Array<Version>,
): Array<Version> {
  const fromLs = window.localStorage.getItem(key);

  if (fromLs) {
    try {
      let result: Array<Version> = JSON.parse(fromLs);

      if (!isExpectedFormat(result)) {
        // Known versions can just be downloaded again.
        if (key === VersionKeys.known) {
          throw new Error(
            `Electron versions in LS does not match expected format`,
          );
        }

        // Local versions are a bit more tricky and might be in an old format (pre 0.5)
        result = migrateVersions(result);
        saveLocalVersions(result);
      }

      return result;
    } catch (error) {
      console.warn(
        `Parsing local Electron versions failed, returning fallback method.`,
      );
    }
  }

  return fallbackMethod();
}

/**
 * Save an array of GitHubVersions to localStorage.
 *
 * @param {VersionKeys} key
 * @param {Array<Version} versions
 */
function saveVersions(key: VersionKeys, versions: Array<Version>) {
  const stringified = JSON.stringify(versions);
  window.localStorage.setItem(key, stringified);
}

/**
 * Gets the current state of a specific version
 * Valid local electron builds are marked as `installed`
 *
 * @param {Version} ver
 * @returns {InstallState}
 */
export function getVersionState(ver: Version): InstallState {
  const { localPath } = ver;
  if (localPath !== undefined) {
    const dir = Installer.getExecPath(localPath);
    if (fs.existsSync(dir)) {
      return InstallState.installed;
    }
  }

  return InstallState.missing;
}

export function makeRunnable(ver: Version): RunnableVersion {
  const ret: RunnableVersion = {
    ...ver,
    version: normalizeVersion(ver.version),
    source: Boolean(ver.localPath) ? VersionSource.local : VersionSource.remote,
    state: getVersionState(ver),
  };

  return ret;
}

/**
 * Return both known as well as local versions.
 *
 * @returns {Array<RunnableVersion>}
 */
export function getElectronVersions(): Array<RunnableVersion> {
  const versions = [...getReleasedVersions(), ...getLocalVersions()];
  return versions.map((ver) => makeRunnable(ver));
}

/**
 * Add a version to the local versions
 *
 * @param {Version} input
 * @returns {Array<Version>}
 */
export function addLocalVersion(input: Version): Array<Version> {
  const versions = getLocalVersions();

  if (!versions.find((v) => v.localPath === input.localPath)) {
    versions.push(input);
  }

  saveLocalVersions(versions);

  return versions;
}

/**
 * Get the Version (if any) that is located at localPath.
 *
 * @param {string} folderPath
 * @returns {Version | undefined}
 */
export function getLocalVersionForPath(
  folderPath: string,
): Version | undefined {
  return getLocalVersions().find((v) => v.localPath === folderPath);
}

/**
 * Retrieves local Electron versions, configured by the user.
 *
 * @returns {Array<Version>}
 */
export function getLocalVersions(): Array<Version> {
  const versions = getVersions(VersionKeys.local, () => []);

  return versions;
}

/**
 * Saves local versions to localStorage.
 *
 * @param {Array<Version | RunnableVersion>} versions
 */
export function saveLocalVersions(versions: Array<Version | RunnableVersion>) {
  const filteredVersions = versions.filter((v) => {
    if (isElectronVersion(v)) {
      return v.source === VersionSource.local;
    }

    return true;
  });

  return saveVersions(VersionKeys.local, filteredVersions);
}

/**
 * Retrieves our best guess regarding the latest Electron versions. Tries to
 * fetch them from localStorage, then from a static releases.json file.
 *
 * @returns {Array<Version>}
 */
function getReleasedVersions(): Array<Version> {
  return getVersions(VersionKeys.known, () => {
    return releasesJSON as Array<Version>;
  });
}

/**
 * Helper to check if this version is from a released major branch.
 *
 * This way when we have a local version of Electron like '999.0.0'
 * we'll know to not try & download 999-x-y.zip from GitHub :D
 *
 * @param {number} major - Electron major version number
 * @returns {boolean} true if there are releases with that major version
 */
export function isReleasedMajor(major: number) {
  const prefix = `${major}.`;
  return getReleasedVersions().some((ver) => ver.version.startsWith(prefix));
}

/**
 * Saves known versions to localStorage.
 *
 * @param {Array<Version>} versions
 */
function saveKnownVersions(versions: Array<Version>) {
  return saveVersions(VersionKeys.known, versions);
}

/**
 * Fetch a list of released versions from electronjs.org.
 *
 * @returns {Promise<Version[]>}
 */
export async function fetchVersions(): Promise<Version[]> {
  const url = 'https://releases.electronjs.org/releases.json';
  const response = await window.fetch(url, {
    cache: 'no-store',
  });
  const data = (await response.json()) as { version: string }[];

  const versions: Version[] = data
    // Don't support anything older than 0.30 (Aug 2015).
    // The oldest version known to releases.json.org is 0.20,
    // Pre-0.24.0 versions were technically 'atom-shell' and cannot
    // be downloaded with @electron/get.
    .filter((ver) => !ver.version.startsWith('0.2'))
    .map(({ version }) => ({ version }));

  console.log(`Fetched ${versions.length} new Electron versions`);
  if (versions.length > 0) saveKnownVersions(versions);
  return versions;
}

/**
 * Is the given array an array of versions?
 *
 * @param {Array<any>} input
 * @returns {boolean}
 */
function isExpectedFormat(input: Array<any>): boolean {
  return input.every((entry) => !!entry.version);
}

/**
 * Migrates old versions, if necessary
 *
 * @param {Array<any>} input
 * @returns {Array<Version>}
 */
function migrateVersions(input: Array<any>): Array<Version> {
  return input
    .filter((item) => !!item)
    .map((item) => {
      const { tag_name, name, url } = item;

      if (!tag_name || !name || !url) return null;

      return {
        version: tag_name,
        name,
        localPath: url,
      };
    })
    .filter((item) => !!item) as Array<Version>;
}

function isElectronVersion(
  input: Version | RunnableVersion,
): input is RunnableVersion {
  return (input as RunnableVersion).source !== undefined;
}

export function getOldestSupportedMajor(): number | undefined {
  const NUM_BRANCHES = parseInt(process.env.NUM_STABLE_BRANCHES || '') || 4;

  return getReleasedVersions()
    .filter((ver) => ver.version.endsWith('.0.0'))
    .map((ver) => Number.parseInt(ver.version))
    .sort((a, b) => a - b)
    .slice(-NUM_BRANCHES)
    .shift();
}
