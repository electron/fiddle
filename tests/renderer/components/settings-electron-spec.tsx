import * as React from 'react';

import { InstallState } from '@vertedinde/fiddle-core';
import { mount, shallow } from 'enzyme';

import {
  ElectronReleaseChannel,
  RunnableVersion,
  VersionSource,
} from '../../../src/interfaces';
import { ElectronSettings } from '../../../src/renderer/components/settings-electron';
import * as versions from '../../../src/renderer/versions';
import { disableDownload } from '../../../src/utils/disable-download';
import { StateMock, VersionsMock } from '../../mocks/mocks';

jest.mock('../../../src/utils/disable-download.ts');

describe('ElectronSettings component', () => {
  let store: StateMock;
  let mockVersions: Record<string, RunnableVersion>;
  let mockVersionsArray: RunnableVersion[];

  beforeEach(() => {
    ({ mockVersions, mockVersionsArray } = new VersionsMock());
    ({ state: store } = (window as any).ElectronFiddle.app);

    store.initVersions('2.0.1', { ...mockVersions });
    store.channelsToShow = [
      ElectronReleaseChannel.stable,
      ElectronReleaseChannel.beta,
    ];

    // Render all the states
    let i = 0;
    store.versionsToShow[i++].state = InstallState.installed;
    store.versionsToShow[i++].state = InstallState.downloading;
    store.versionsToShow[i++].state = InstallState.missing;
    store.versionsToShow[i++].state = InstallState.installing;
  });

  it('renders', () => {
    const spy = jest
      .spyOn(versions, 'getOldestSupportedMajor')
      .mockReturnValue(9);

    const moreVersions: RunnableVersion[] = [
      {
        source: VersionSource.local,
        state: InstallState.installed,
        version: '3.0.0',
      },
      {
        source: VersionSource.remote,
        state: InstallState.installed,
        version: '3.0.0-nightly.1',
      },
    ];

    for (const ver of moreVersions) {
      store.versions[ver.version] = ver;
      store.versionsToShow.unshift(ver);
    }

    const wrapper = shallow(<ElectronSettings appState={store as any} />);
    expect(wrapper).toMatchSnapshot();

    spy.mockRestore();
  });

  it('handles removing a version', async () => {
    store.versions['3.0.0-nightly.1'] = {
      state: InstallState.installed,
      version: '3.0.0-nightly.1',
      source: VersionSource.local,
    };

    store.versions['3.0.0'] = {
      state: InstallState.installed,
      version: '3.0.0',
      source: VersionSource.local,
    };

    const wrapper = mount(<ElectronSettings appState={store as any} />);

    wrapper
      .find('.electron-versions-table .bp3-button')
      .first()
      .simulate('click');

    expect(store.removeVersion).toHaveBeenCalledTimes(1);
  });

  it('handles downloading a version', async () => {
    const version = '3.0.0';
    const ver = {
      source: VersionSource.remote,
      state: InstallState.missing,
      version,
    };
    store.versions = { version: ver };
    store.versionsToShow = [ver];

    const wrapper = mount(<ElectronSettings appState={store as any} />);

    wrapper
      .find('.electron-versions-table .bp3-button')
      .first()
      .simulate('click');

    expect(store.downloadVersion).toHaveBeenCalledTimes(1);
  });

  it('handles missing local versions', () => {
    const version = '99999.0.0';
    const ver = {
      source: VersionSource.local,
      state: InstallState.missing,
      version,
    };
    store.versions = { version: ver };
    store.versionsToShow = [ver];

    const wrapper = mount(<ElectronSettings appState={store as any} />);

    wrapper
      .find('.electron-versions-table .bp3-button')
      .first()
      .simulate('click');

    expect(store.removeVersion).toHaveBeenCalledTimes(1);
  });

  it('handles the deleteAll()', async () => {
    const wrapper = shallow(<ElectronSettings appState={store as any} />);
    const instance = wrapper.instance() as any;
    await instance.handleDeleteAll();

    expect(store.removeVersion).toHaveBeenCalledTimes(mockVersionsArray.length);
  });

  it('handles the downloadAll()', async () => {
    const wrapper = shallow(<ElectronSettings appState={store as any} />);
    const instance = wrapper.instance() as any;
    await instance.handleDownloadAll();

    expect(store.downloadVersion).toHaveBeenCalled();
  });

  describe('handleDownloadClick()', () => {
    it('kicks off an update of Electron versions', async () => {
      const wrapper = shallow(<ElectronSettings appState={store as any} />);
      const instance = wrapper.instance() as any;
      await instance.handleDownloadClick();

      expect(store.updateElectronVersions).toHaveBeenCalled();
    });
  });

  describe('handleAddVersion()', () => {
    it('toggles the add version dialog', () => {
      const wrapper = shallow(<ElectronSettings appState={store as any} />);
      const instance = wrapper.instance() as any;
      instance.handleAddVersion();

      expect(store.toggleAddVersionDialog).toHaveBeenCalled();
    });
  });

  describe('handleStateChange()', () => {
    it('toggles remote versions', async () => {
      const id = 'showUndownloadedVersions';
      for (const checked of [true, false]) {
        const wrapper = shallow(<ElectronSettings appState={store as any} />);
        const instance = wrapper.instance() as any;
        await instance.handleStateChange({
          currentTarget: { checked, id },
        });
        expect(store[id]).toBe(checked);
      }
    });
  });

  describe('handleShowObsoleteChange()', () => {
    it('toggles obsolete versions', async () => {
      const id = 'showObsoleteVersions';
      for (const checked of [true, false]) {
        const wrapper = shallow(<ElectronSettings appState={store as any} />);
        const instance = wrapper.instance() as any;
        await instance.handleShowObsoleteChange({
          currentTarget: { checked, id },
        });
        expect(store[id]).toBe(checked);
      }
    });
  });

  describe('handleChannelChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(<ElectronSettings appState={store as any} />);
      store.showChannels.mockImplementation((ids: ElectronReleaseChannel[]) =>
        store.channelsToShow.push(...ids),
      );
      store.hideChannels.mockImplementation(
        (ids: ElectronReleaseChannel[]) =>
          (store.channelsToShow = store.channelsToShow.filter(
            (id: ElectronReleaseChannel) => !ids.includes(id),
          )),
      );
      const instance = wrapper.instance() as any;
      await instance.handleChannelChange({
        currentTarget: {
          id: ElectronReleaseChannel.stable,
          checked: false,
        },
      });

      await instance.handleChannelChange({
        currentTarget: {
          id: ElectronReleaseChannel.nightly,
          checked: true,
        },
      });

      expect(store.channelsToShow).toEqual([
        ElectronReleaseChannel.beta,
        ElectronReleaseChannel.nightly,
      ]);
    });
  });

  describe('disableDownload()', () => {
    it('disables download buttons where return values are true', () => {
      (disableDownload as any).mockReturnValue(true);

      const version = '3.0.0';
      const ver = {
        source: VersionSource.remote,
        state: InstallState.missing,
        version,
      };

      store.versions = { version: ver };

      store.versionsToShow = [ver, { ...ver }, { ...ver }];

      const wrapper = shallow(<ElectronSettings appState={store as any} />);

      expect(wrapper.find('.disabled-version')).toHaveLength(3);
    });
  });
});
