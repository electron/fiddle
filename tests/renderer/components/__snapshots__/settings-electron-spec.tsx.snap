// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElectronSettings component renders 1`] = `
<div
  className="settings-electron"
>
  <h1>
    Electron Settings
  </h1>
  <Blueprint3.Callout>
    <Blueprint3.FormGroup
      label="Include Electron versions:"
    >
      <Blueprint3.Tooltip
        content="Can't disable channel of selected version (2.0.1)"
        disabled={false}
        hoverCloseDelay={0}
        hoverOpenDelay={100}
        intent="primary"
        key="Stable"
        minimal={false}
        position="bottom"
        transitionDuration={100}
      >
        <Blueprint3.Checkbox
          checked={true}
          disabled={true}
          id="Stable"
          inline={true}
          label="Stable"
          onChange={[Function]}
        />
      </Blueprint3.Tooltip>
      <Blueprint3.Tooltip
        content="Can't disable channel of selected version (2.0.1)"
        disabled={true}
        hoverCloseDelay={0}
        hoverOpenDelay={100}
        intent="primary"
        key="Beta"
        minimal={false}
        position="bottom"
        transitionDuration={100}
      >
        <Blueprint3.Checkbox
          checked={true}
          disabled={false}
          id="Beta"
          inline={true}
          label="Beta"
          onChange={[Function]}
        />
      </Blueprint3.Tooltip>
      <Blueprint3.Tooltip
        content="Can't disable channel of selected version (2.0.1)"
        disabled={true}
        hoverCloseDelay={0}
        hoverOpenDelay={100}
        intent="primary"
        key="Nightly"
        minimal={false}
        position="bottom"
        transitionDuration={100}
      >
        <Blueprint3.Checkbox
          checked={false}
          disabled={false}
          id="Nightly"
          inline={true}
          label="Nightly"
          onChange={[Function]}
        />
      </Blueprint3.Tooltip>
      <Blueprint3.Checkbox
        checked={false}
        id="showUndownloadedVersions"
        inline={true}
        label="Not downloaded"
        onChange={[Function]}
      />
      <Blueprint3.Tooltip
        content="Include versions that have reached end-of-life (older than 9.0.0)"
        hoverCloseDelay={0}
        hoverOpenDelay={100}
        intent="primary"
        minimal={false}
        position="bottom"
        transitionDuration={100}
      >
        <Blueprint3.Checkbox
          checked={false}
          id="showObsoleteVersions"
          inline={true}
          label="Obsolete"
          onChange={[Function]}
        />
      </Blueprint3.Tooltip>
    </Blueprint3.FormGroup>
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.ButtonGroup
      fill={true}
    >
      <Blueprint3.Button
        icon="numbered-list"
        onClick={[Function]}
        text="Update Electron Release List"
      />
      <Blueprint3.Button
        disabled={false}
        icon="download"
        onClick={[Function]}
        text="Download All Versions"
      />
      <Blueprint3.Button
        disabled={false}
        icon="trash"
        onClick={[Function]}
        text="Delete All Downloads"
      />
      <Blueprint3.Button
        icon="document-open"
        onClick={[Function]}
        text="Add Local Electron Build"
      />
    </Blueprint3.ButtonGroup>
    <HTMLTable
      className="electron-versions-table"
      striped={true}
    >
      <thead>
        <tr>
          <th>
            Version
          </th>
          <th>
            Status
          </th>
          <th
            className="action"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="3.0.0-nightly.1"
        >
          <td>
            3.0.0-nightly.1
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              fill={true}
              icon="trash"
              onClick={[Function]}
              small={true}
              text="Delete"
            />
          </td>
        </tr>
        <tr
          key="3.0.0"
        >
          <td>
            3.0.0
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              fill={true}
              icon="trash"
              onClick={[Function]}
              small={true}
              text="Remove"
            />
          </td>
        </tr>
        <tr
          key="2.0.3"
        >
          <td>
            2.0.3
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              fill={true}
              icon="trash"
              onClick={[Function]}
              small={true}
              text="Delete"
            />
          </td>
        </tr>
        <tr
          key="2.0.2"
        >
          <td>
            2.0.2
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="cloud-download"
              />
               
              Downloading
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              disabled={true}
              fill={true}
              icon="cloud-download"
              loading={true}
              small={true}
              text="Downloading"
            />
          </td>
        </tr>
        <tr
          key="2.0.1"
        >
          <td>
            2.0.1
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="cloud"
              />
               
              Not downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Tooltip
              content="Can't remove currently active Electron version (2.0.1)"
              hoverCloseDelay={0}
              hoverOpenDelay={100}
              intent="primary"
              minimal={false}
              position="auto"
              transitionDuration={100}
            >
              <Blueprint3.AnchorButton
                className="disabled-version"
                disabled={true}
                icon="cloud-download"
                text="Download"
              />
            </Blueprint3.Tooltip>
          </td>
        </tr>
        <tr
          key="2.0.0"
        >
          <td>
            2.0.0
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              disabled={true}
              fill={true}
              icon="cloud-download"
              loading={true}
              small={true}
              text="Downloading"
            />
          </td>
        </tr>
        <tr
          key="1.8.7"
        >
          <td>
            1.8.7
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              fill={true}
              icon="trash"
              onClick={[Function]}
              small={true}
              text="Delete"
            />
          </td>
        </tr>
      </tbody>
    </HTMLTable>
  </Blueprint3.Callout>
</div>
`;
